File: jbake-core/src/main/java/org/jbake/app/Crawler.java
package org.jbake.app;

import org.apache.commons.configuration2.CompositeConfiguration;
import org.apache.commons.io.FilenameUtils;
import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.app.configuration.JBakeConfigurationFactory;
import org.jbake.model.DocumentModel;
import org.jbake.model.DocumentStatus;
import org.jbake.model.DocumentTypes;
import org.jbake.model.ModelAttributes;
import org.jbake.util.HtmlUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.net.*;
import java.nio.charset.*;
import java.util.*;

/**
 * Crawls a file system looking for content.
 *
 * @author Jonathan Bullock <a href="mailto:jonbullock@gmail.com">jonbullock@gmail.com</a>
 */
public class Crawler {

    private static final Logger logger = LoggerFactory.getLogger(Crawler.class);
    private final ContentStore db;
    private final JBakeConfiguration config;
    private final Parser parser;

    /**
     * @param db     Database instance for content
     * @param source Base directory where content directory is located
     * @param config Project configuration
     * @deprecated Use {@link #Crawler(ContentStore, JBakeConfiguration)} instead.
     * <p>
     * Creates new instance of Crawler.
     */
    @Deprecated
    public Crawler(ContentStore db, File source, CompositeConfiguration config) {
        this.db = db;
        this.config = new JBakeConfigurationFactory().createDefaultJbakeConfiguration(source, config);
        this.parser = new Parser(this.config);
    }

    /**
     * Creates new instance of Crawler.
     *
     * @param db     Database instance for content
     * @param config Project configuration
     */
    public Crawler(ContentStore db, JBakeConfiguration config) {
        this.db = db;
        this.config = config;
        this.parser = new Parser(config);
    }

    public void crawl() {
        crawl(config.getContentFolder());

        logger.info("Content detected:");
        for (String docType : DocumentTypes.getDocumentTypes()) {
            long count = db.getDocumentCount(docType);
            if (count > 0) {
                logger.info("Parsed {} files of type: {}", count, docType);
            }
        }
    }

    public void crawlDataFiles() {
        crawlDataFiles(config.getDataFolder());

        logger.info("Data files detected:");
        String docType = config.getDataFileDocType();
        long count = db.getDocumentCount(docType);
        if (count > 0) {
            logger.info("Parsed {} files", count);
        }
    }

    /**
     * Crawl all files and folders looking for content.
     *
     * @param path Folder to start from
     */
    private void crawl(File path) {
        File[] contents = path.listFiles(FileUtil.getFileFilter(config));
        if (contents != null) {
            Arrays.sort(contents);
            for (File sourceFile : contents) {
                if (sourceFile.isFile()) {
                    crawlFile(sourceFile);
                } else if (sourceFile.isDirectory()) {
                    crawl(sourceFile);
                }
            }
        }
    }

    private void crawlFile(File sourceFile) {

        StringBuilder sb = new StringBuilder();
        sb.append("Processing [").append(sourceFile.getPath()).append("]... ");
        String sha1 = buildHash(sourceFile);
        String uri = buildURI(sourceFile);
        DocumentStatus status = findDocumentStatus(uri, sha1);
        if (status == DocumentStatus.UPDATED) {
            sb.append(" : modified ");
            db.deleteContent(uri);
        } else if (status == DocumentStatus.IDENTICAL) {
            sb.append(" : same ");
        } else if (DocumentStatus.NEW == status) {
            sb.append(" : new ");
        }

        logger.info("{}", sb);

        if (status != DocumentStatus.IDENTICAL) {
            processSourceFile(sourceFile, sha1, uri);
        }
    }

    /**
     * Crawl all files and folders looking for data files.
     *
     * @param path Folder to start from
     */
    private void crawlDataFiles(File path) {
        File[] contents = path.listFiles(FileUtil.getDataFileFilter());
        if (contents != null) {
            Arrays.sort(contents);
            for (File sourceFile : contents) {
                if (sourceFile.isFile()) {
                    StringBuilder sb = new StringBuilder();
                    sb.append("Processing [").append(sourceFile.getPath()).append("]... ");
                    String sha1 = buildHash(sourceFile);
                    String uri = buildDataFileURI(sourceFile);
                    boolean process = true;
                    DocumentStatus status = DocumentStatus.NEW;
                    String docType = config.getDataFileDocType();
                    status = findDocumentStatus(uri, sha1);
                    if (status == DocumentStatus.UPDATED) {
                        sb.append(" : modified ");
                        db.deleteContent(uri);
                    } else if (status == DocumentStatus.IDENTICAL) {
                        sb.append(" : same ");
                        process = false;
                    }
                    if (!process) {
                        break;
                    }
                    if (DocumentStatus.NEW == status) {
                        sb.append(" : new ");
                    }
                    if (process) { // new or updated
                        crawlDataFile(sourceFile, sha1, uri, docType);
                    }
                    logger.info("{}", sb);
                }
                if (sourceFile.isDirectory()) {
                    crawlDataFiles(sourceFile);
                }
            }
        }
    }

    private String buildHash(final File sourceFile) {
        String sha1;
        try {
            sha1 = FileUtil.sha1(sourceFile);
        } catch (Exception e) {
            logger.error("unable to build sha1 hash for source file '{}'", sourceFile);
            sha1 = "";
        }
        return sha1;
    }

    private String buildURI(final File sourceFile) {
        String uri = FileUtil.asPath(sourceFile).replace(FileUtil.asPath(config.getContentFolder()), "");

        if (useNoExtensionUri(uri)) {
            // convert URI from xxx.html to xxx/index.html
            uri = createNoExtensionUri(uri);
        } else {
            uri = createUri(uri);
        }

        // strip off leading / to enable generating non-root based sites
        if (uri.startsWith(FileUtil.URI_SEPARATOR_CHAR)) {
            uri = uri.substring(1);
        }

        return uri;
    }

    private String buildDataFileURI(final File sourceFile) {
        String uri = FileUtil.asPath(sourceFile).replace(FileUtil.asPath(config.getDataFolder()), "");
        // strip off leading /
        if (uri.startsWith(FileUtil.URI_SEPARATOR_CHAR)) {
            uri = uri.substring(1, uri.length());
        }
        return uri;
    }

    // TODO: Refactor - parametrize the following two methods into one.
    // commons-codec's URLCodec could be used when we add that dependency.
    private String createUri(String uri) {
        try {
            return FileUtil.URI_SEPARATOR_CHAR
                + FilenameUtils.getPath(uri)
                + URLEncoder.encode(FilenameUtils.getBaseName(uri), StandardCharsets.UTF_8.name())
                + config.getOutputExtension();
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException("Missing UTF-8 encoding??", e); // Won't happen unless JDK is broken.
        }
    }

    private String createNoExtensionUri(String uri) {
        try {
            return FileUtil.URI_SEPARATOR_CHAR
                + FilenameUtils.getPath(uri)
                + URLEncoder.encode(FilenameUtils.getBaseName(uri), StandardCharsets.UTF_8.name())
                + FileUtil.URI_SEPARATOR_CHAR
                + "index"
                + config.getOutputExtension();
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException("Missing UTF-8 encoding??", e); // Won't happen unless JDK is broken.
        }
    }

    private boolean useNoExtensionUri(String uri) {
        boolean noExtensionUri = config.getUriWithoutExtension();
        String noExtensionUriPrefix = config.getPrefixForUriWithoutExtension();

        return noExtensionUri
            && (noExtensionUriPrefix != null)
            && (noExtensionUriPrefix.length() > 0)
            && uri.startsWith(noExtensionUriPrefix);
    }

    private void crawlDataFile(final File sourceFile, final String sha1, final String uri, final String documentType) {
        try {
            DocumentModel document = parser.processFile(sourceFile);
            if (document != null) {
                document.setSha1(sha1);
                document.setRendered(true);
                document.setFile(sourceFile.getPath());
                document.setSourceUri(uri);
                document.setType(documentType);

                db.addDocument(document);
            } else {
                logger.warn("{} couldn't be parsed so it has been ignored!", sourceFile);
            }
        } catch (Exception ex) {
            throw new RuntimeException("Failed crawling file: " + sourceFile.getPath() + " " + ex.getMessage(), ex);
        }
    }

    private void processSourceFile(final File sourceFile, final String sha1, final String uri) {
        DocumentModel document = parser.processFile(sourceFile);

        if (document != null) {
            if (DocumentTypes.contains(document.getType())) {
                addAdditionalDocumentAttributes(document, sourceFile, sha1, uri);

                if (config.getImgPathUpdate()) {
                    // Prevent image source url's from breaking
                    HtmlUtil.fixImageSourceUrls(document, config);
                }

                db.addDocument(document);
            } else {
                logger.warn("{} has an unknown document type '{}' and has been ignored!", sourceFile, document.getType());
            }
        } else {
            logger.warn("{} has an invalid header, it has been ignored!", sourceFile);
        }
    }

    private void addAdditionalDocumentAttributes(DocumentModel document, File sourceFile, String sha1, String uri) {
        document.setRootPath(getPathToRoot(sourceFile));
        document.setSha1(sha1);
        document.setRendered(false);
        document.setFile(sourceFile.getPath());
        document.setSourceUri(uri);
        document.setUri(uri);
        document.setCached(true);

        if (document.getStatus().equals(ModelAttributes.Status.PUBLISHED_DATE)
                && (document.getDate() != null)
                && new Date().after(document.getDate())) {
            document.setStatus(ModelAttributes.Status.PUBLISHED);
        }

        if (config.getUriWithoutExtension()) {
            document.setNoExtensionUri(uri.replace("/index.html", "/"));
        }
    }

    private String getPathToRoot(File sourceFile) {
        return FileUtil.getUriPathToContentRoot(config, sourceFile);
    }

    private DocumentStatus findDocumentStatus(String uri, String sha1) {
        DocumentList<DocumentModel> match = db.getDocumentStatus(uri);
        if (!match.isEmpty()) {
            DocumentModel document = match.get(0);
            String oldHash = document.getSha1();
            if (!oldHash.equals(sha1) || !document.getRendered()) {
                return DocumentStatus.UPDATED;
            } else {
                return DocumentStatus.IDENTICAL;
            }
        } else {
            return DocumentStatus.NEW;
        }
    }

}

File: jbake-core/src/main/java/org/jbake/app/Crawler.kt
package org.jbake.app

import org.apache.commons.io.FilenameUtils
import org.jbake.app.FileUtil.URI_SEPARATOR_CHAR
import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.model.DocumentModel
import org.jbake.model.DocumentTypeRegistry
import org.jbake.model.ModelAttributes
import org.jbake.util.HtmlUtil
import org.jbake.util.Logging.logger
import org.slf4j.Logger
import java.io.File
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * Crawls a file system looking for content.
 */
class Crawler {

    private val db: ContentStore
    private val config: JBakeConfiguration
    private val parser: Parser


    constructor(db: ContentStore, config: JBakeConfiguration) {
        this.db = db
        this.config = config
        this.parser = Parser(this.config)
    }

    fun crawlContentDirectory() {
        crawlDirectory(config.contentDir)

        for (docType in DocumentTypeRegistry.documentTypes) {
            val count = db.getDocumentCount(docType)
            if (count > 0) log.info("Parsed {} files of type: {}", count, docType)
        }
    }

    /**
     * Crawl all files and directorys looking for content.
     */
    private fun crawlDirectory(startFromDirectory: File) {
        val contents = startFromDirectory.listFiles(FileUtil.getFileFilter(config)) ?: return

        Arrays.sort(contents)
        for (sourceFile in contents) {
            when {
                sourceFile.isFile -> crawlFile(sourceFile)
                sourceFile.isDirectory -> crawlDirectory(sourceFile)
            }
        }
    }

    private fun crawlFile(sourceFile: File) {
        val sha1 = runCatching { FileUtil.sha1(sourceFile) }
            .getOrElse { log.error("Unable to build SHA1 hash for source file '$sourceFile'"); "" }

        val uri = buildURI(sourceFile)

        when (findDocumentStatus(uri, sha1)) {
            DocumentStatus.UPDATED -> { log.info("MODIFIED:" + sourceFile.path); db.deleteContent(uri) }
            DocumentStatus.IDENTICAL -> log.info("SAME:    " + sourceFile.path)
            DocumentStatus.NEW -> log.info("NEW:     " + sourceFile.path)
        }

        if (findDocumentStatus(uri, sha1) != DocumentStatus.IDENTICAL)
            processSourceFile(sourceFile, sha1, uri)
    }

    /**
     * Crawl all files and directorys looking for data files.
     */
    fun crawlDataFiles() {
        crawlDataFiles(config.dataDir)

        val count = db.getDocumentCount(config.dataFileDocType)
        if (count > 0) log.info("Parsed {} files", count)
    }

    private fun crawlDataFiles(startFromDirectory: File) {
        val contents = startFromDirectory.listFiles(FileUtil.dataFileFilter) ?: return

        for (sourceFile in contents.sorted()) {
            if (sourceFile.isDirectory) {
                crawlDataFiles(sourceFile)
                continue
            }

            val sha1 = runCatching { FileUtil.sha1(sourceFile) }
                .getOrElse { log.error("Unable to build SHA1 hash for source file '$sourceFile'"); "" }

            val uri = buildDataFileUri(sourceFile)

            when (findDocumentStatus(uri, sha1)) {
                DocumentStatus.UPDATED -> { log.info("MODIFIED:" + sourceFile.path); db.deleteContent(uri) }
                DocumentStatus.IDENTICAL -> log.info("SAME:    " + sourceFile.path).also { continue }
                DocumentStatus.NEW -> log.info("NEW:     " + sourceFile.path)
            }

            crawlDataFile(sourceFile, sha1, uri, config.dataFileDocType)
        }
    }

    private fun buildURI(sourceFile: File): String {
        val uri = FileUtil.asPath(sourceFile).replace(FileUtil.asPath(config.contentDir), "")

        val processedUri =
            // Convert URI from xxx.html to xxx/index.html .
            if (useNoExtensionUri(uri))
                createNoExtensionUri(uri)
            else createUri(uri)

        // Strip off leading / to enable generating non-root based sites.
        return processedUri.removePrefix(URI_SEPARATOR_CHAR)
    }

    private fun buildDataFileUri(sourceFile: File): String {
        return FileUtil.asPath(sourceFile)
            .replace(FileUtil.asPath(config.dataDir), "")
            .removePrefix(URI_SEPARATOR_CHAR)
    }

    // TODO: Refactor - parametrize the following two methods into one. commons-codec's URLCodec could be used when we add that dependency.
    private fun createUri(uri: String): String {
        return (URI_SEPARATOR_CHAR
                + FilenameUtils.getPath(uri)
                + URLEncoder.encode(FilenameUtils.getBaseName(uri), StandardCharsets.UTF_8.name())
                + config.outputExtension)
    }

    private fun createNoExtensionUri(uri: String): String {
        return (URI_SEPARATOR_CHAR
                + FilenameUtils.getPath(uri)
                + URLEncoder.encode(FilenameUtils.getBaseName(uri), StandardCharsets.UTF_8.name())
                + URI_SEPARATOR_CHAR
                + "index"
                + config.outputExtension)
    }

    private fun useNoExtensionUri(uri: String): Boolean {
        val noExtensionUri = config.uriWithoutExtension
        val noExtensionUriPrefix = config.prefixForUriWithoutExtension

        return noExtensionUri
                && noExtensionUriPrefix != null
                && noExtensionUriPrefix.isNotEmpty()
                && uri.startsWith(noExtensionUriPrefix)
    }

    private fun crawlDataFile(sourceFile: File, sha1: String, uri: String, documentType: String) {
        try {
            val document = parser.processFile(sourceFile) ?: run {
                log.warn("{} couldn't be parsed so it has been ignored!", sourceFile)
                return
            }

            document.sha1 = sha1
            document.rendered = true
            document.file = sourceFile.path
            document.sourceUri = uri
            document.type = documentType

            db.addDocument(document)
        } catch (ex: Exception) {
            throw RuntimeException("Failed crawling file: " + sourceFile.path + " " + ex.message, ex)
        }
    }

    private fun processSourceFile(sourceFile: File, sha1: String, uri: String) {
        val document = parser.processFile(sourceFile) ?: run {
            log.warn("{} has an invalid header, it has been ignored!", sourceFile)
            return
        }

        if (!DocumentTypeRegistry.contains(document.type)) {
            log.warn("{} has an unknown document type '{}' and has been ignored!", sourceFile, document.type)
            return
        }

        addAdditionalDocumentAttributes(document, sourceFile, sha1, uri)

        // Prevent image source URL's from breaking.
        if (config.imgPathUpdate)
            HtmlUtil.fixImageSourceUrls(document, config)

        db.addDocument(document)
        db.addDocument(document)
    }

    private fun addAdditionalDocumentAttributes(document: DocumentModel, sourceFile: File, sha1: String, uri: String) {
        document.rootPath = FileUtil.getUriPathToSourceRoot(config, sourceFile)
        document.sha1 = sha1
        document.rendered = false
        document.file = sourceFile.path
        document.sourceUri = uri
        document.uri = uri
        document.cached = true

        if (document.status == ModelAttributes.Status.PUBLISHED_DATE
                && document.date != null
                && Date().after(document.date)
        )
            document.status = ModelAttributes.Status.PUBLISHED

        if (config.uriWithoutExtension)
            document.noExtensionUri = uri.replace("/index.html", "/")
    }

    private fun findDocumentStatus(uri: String, sha1: String): DocumentStatus {
        val match = db.getDocumentStatus(uri)
        if (match.isEmpty()) return DocumentStatus.NEW

        val document = match[0]
        val oldHash = document.sha1

        return if (oldHash != sha1 || !document.rendered) DocumentStatus.UPDATED
            else DocumentStatus.IDENTICAL
    }

    private val log: Logger by logger()
}


/** Enumeration used to determine whether rendering of a document should be done. */
enum class DocumentStatus {
    NEW,
    UPDATED,
    IDENTICAL
}
